File: src/nallar/ps2edit/ui/MessageConsole.java
package nallar.ps2edit.ui;

import javax.swing.text.*;
import java.awt.*;
import java.io.*;

public class MessageConsole {
	private JTextComponent textComponent;
	private Document document;
	private boolean isAppend;

	public MessageConsole(JTextComponent textComponent) {
		this(textComponent, true);
	}

	public MessageConsole(JTextComponent textComponent, boolean isAppend) {
		this.textComponent = textComponent;
		this.document = textComponent.getDocument();
		this.isAppend = isAppend;
		textComponent.setEditable(false);

		ConsoleOutputStream cos = new ConsoleOutputStream(Color.black, System.out);
		System.setOut(new PrintStream(cos, true));
		cos = new ConsoleOutputStream(Color.red, System.err);
		System.setErr(new PrintStream(cos, true));
	}

	private class ConsoleOutputStream extends ByteArrayOutputStream {
		private final String EOL = System.getProperty("line.separator");
		private SimpleAttributeSet attributes;
		private PrintStream printStream;
		private StringBuffer buffer = new StringBuffer(80);
		private boolean isFirstLine;

		public ConsoleOutputStream(Color textColor, PrintStream printStream) {
			if (textColor != null) {
				attributes = new SimpleAttributeSet();
				StyleConstants.setForeground(attributes, textColor);
			}

			this.printStream = printStream;

			if (isAppend)
				isFirstLine = true;
		}

		public void flush() {
			String message = toString();

			if (message.length() == 0) return;

			if (isAppend)
				handleAppend(message);
			else
				handleInsert(message);

			reset();
		}

		private void handleAppend(String message) {
			if (document.getLength() == 0)
				buffer.setLength(0);

			if (EOL.equals(message)) {
				buffer.append(message);
			} else {
				buffer.append(message);
				clearBuffer();
			}

		}

		private void handleInsert(String message) {
			buffer.append(message);

			if (EOL.equals(message)) {
				clearBuffer();
			}
		}

		private void clearBuffer() {
			if (isFirstLine && document.getLength() != 0) {
				buffer.insert(0, "\n");
			}

			isFirstLine = false;
			String line = buffer.toString();

			try {
				if (isAppend) {
					int offset = document.getLength();
					document.insertString(offset, line, attributes);
					textComponent.setCaretPosition(document.getLength());
				} else {
					document.insertString(0, line, attributes);
					textComponent.setCaretPosition(0);
				}
			} catch (BadLocationException ignored) {
			}

			if (printStream != null) {
				printStream.print(line);
			}

			buffer.setLength(0);
		}
	}
}

File: src/main/kotlin/net/ps2stats/ui/MessageConsole.kt
package net.ps2stats.ui

import java.awt.Color
import java.io.ByteArrayOutputStream
import java.io.PrintStream
import javax.swing.text.*

class MessageConsole @JvmOverloads constructor(private val textComponent: JTextComponent, private val isAppend: Boolean = true) {
	private val document: Document

	init {
		this.document = textComponent.document
		textComponent.isEditable = false

		var cos = ConsoleOutputStream(Color.black, System.out)
		System.setOut(PrintStream(cos, true))
		cos = ConsoleOutputStream(Color.red, System.err)
		System.setErr(PrintStream(cos, true))
	}

	private inner class ConsoleOutputStream(textColor: Color?, private val printStream: PrintStream?) : ByteArrayOutputStream() {
		private val EOL = System.getProperty("line.separator")
		private var attributes: SimpleAttributeSet? = null
		private val buffer = StringBuffer(80)
		private var isFirstLine: Boolean = false

		init {
			if (textColor != null) {
				attributes = SimpleAttributeSet()
				StyleConstants.setForeground(attributes!!, textColor)
			}

			if (isAppend)
				isFirstLine = true
		}

		override fun flush() {
			val message = toString()

			if (message.isEmpty()) return
			if (isAppend)
				handleAppend(message)
			else
				handleInsert(message)
			reset()
		}

		private fun handleAppend(message: String) {
			if (document.length == 0)
				buffer.setLength(0)

			if (EOL == message) {
				buffer.append(message)
			} else {
				buffer.append(message)
				clearBuffer()
			}

		}

		private fun handleInsert(message: String) {
			buffer.append(message)

			if (EOL == message) {
				clearBuffer()
			}
		}

		private fun clearBuffer() {
			if (isFirstLine && document.length != 0) {
				buffer.insert(0, "\n")
			}

			isFirstLine = false
			val line = buffer.toString()

			try {
				if (isAppend) {
					val offset = document.length
					document.insertString(offset, line, attributes)
					textComponent.caretPosition = document.length
				} else {
					document.insertString(0, line, attributes)
					textComponent.caretPosition = 0
				}
			} catch (ignored: BadLocationException) {
			}

			printStream?.print(line)

			buffer.setLength(0)
		}
	}
}
